### PR TITLE
Add batch timeout

### PR DESCRIPTION
**Pull request by jkilpatr**
Batch timeouts allow us to close an edge case where low fee transactions may be stuck in a batch.

For example a user may send 50c of FooToken to Ethereum with 5c of FooToken as the fee. Since it costs $10 in ETH to relay the batch no relayer will do so on their behalf. They will simply have to pony up those funds if they want their FooToken back. 

This patch adds a batch timeout. All batches are produced with an Ethereum block height after which they are no longer valid. MsgDepositClaim and MsgWithdrawClaim ferry the latest Ethereum block height with them. Once a block height on Ethereum higher than the timeout value for the block is observed the transactions can be released and the users tokens refunded.

This means that a batch can time out and not have it's funds released if the bridge is totally inactive. A deposit or any other batch going through will finish the process of releasing funds to the user. This is required to be totally sure of the block height on Ethereum and prevent any possibility of a double spend thanks to Ethereum block speed shenanigans.

The hardest part about this change is figuring out what the *current* block height on Ethereum is when we go to make a batch. The Ethereum block height is only updated when deposits or withdraws occur. For an inactive bridge it may have been days since the last deposit or withdraw! How then do we come up with a batch timeout value?

Two new governance parameters, expressing the average block time for Cosmos and the average block time for Ethereum are included in this patch. Using the last known Ethereum block, the Cosmos block it was recorded at and average block times for both we can estimate the present Ethereum block and add the desired timeout time to it.

For a rough estimate of how accurate (or inaccurate) this is lets do some math and think about some cases.

Ethereum's block speed as only dramatically changed about [4 times in it's history](https://etherscan.io/chart/blocktime). Mostly increases.

If Ethereum is running at half the regular block speed, batches take twice as long to time out. This isn't solvable by any change we could possibly make to our code. So long as we have block based timeouts we have to deal with this possibility. This is also the most likely possibility given the history of Ethereum. 

If Ethereum is running at double speed batches may time out quickly. Or even be created in a state that's already impossible to submit. 

In order to generate batches that are instantly invalid the bridge would have to be totally inactive (no deposits or withdraws) for more five days with blocks being produced 20% faster than normal. The when a deposit occurred things would sync up again. Of course a governance parameter update would also fix things. 

tl;dr

* batches are produced with an *estimated* timeout time, but even in the worst cases the inaccuracy on this estimate isn't bad
* transactions are released from batches only when we are *sure* the timeout height has elapsed.

### CODE REVIEW

File: module/proto/peggy/v1/genesis.proto [hunk 1]
```diff
@@ -34,19 +34,22 @@ message Params {
   uint64 signed_valsets_window            = 6;
   uint64 signed_batches_window            = 7;
   uint64 signed_claims_window             = 8;
-  bytes  slash_fraction_valset            = 9  [
+  uint64 target_batch_timeout_millis      = 10;
```
Comment (jackzampolin) [inline module/proto/peggy/v1/genesis.proto:37]: Lets remove the `target_` and `_millis` from the names here. We can document that these values are in milliseconds.

Comment (jkilpatr) [reply module/proto/peggy/v1/genesis.proto:37]: I think target is important, it's an estimate, not a precise value. Having millis or not having millis I don't really care about.

File: module/x/peggy/abci.go [hunk 2]
```diff
@@ -191,3 +192,22 @@ func attestationTally(ctx sdk.Context, k keeper.Keeper) {
 		}
 	}
 }
+
+// cleanupTimedOutBatches deletes batches that have passed their expiration on Ethereum
+// keep in mind several things when modifying this function
+// A) unlike nonces timeouts are not monotonically increasing, meaning batch 5 can have a later timeout than batch 6
```
Comment (jtremback) [inline module/x/peggy/abci.go:198]: This should never be the case if everything is working correctly.

**Review by jackzampolin — APPROVED** [review]
name changes on the parameter names but otherwise looks good!

**Comment by jtremback** [comment]
I'd say go ahead and merge this, but there are a few issues:

- By using block height we make the implementation more complicated and less predictable. Using timestamps would be a superior user experience in most cases, and would simplify the implementation (no need to track block height like you do here). The tradeoff is that we don't know what's involved in spoofing an Ethereum timestamp. We would need to do some more research and come up with a certain buffer time beyond which we don't think a timestamp can be spoofed. Then we would wait through the buffer before freeing a batch on the Cosmos side. I think that it's probably impossible to spoof a timestamp more than a minute, but I have no idea.
- In the current implementation, the only way the Cosmos module gains knowledge of the current block height is from an event from Peggy. This multiplies any block time inaccuracies, as well as making it so that batches cannot be freed until a new event has come in. An easy upgrade to this would be to have a process separate from the event oracle that brings in block height. However, someone embarking on this journey would first want to check whether the timestamp route could work. 

Can you make issues for these two upgrades?

**Comment by jackzampolin** [comment]
I like the timestamp suggestion

**Comment by jtremback** [comment]
@jackzampolin we would need to find some serious research work on how much timestamps can be spoofed by. I haven't seen any yet.

**Comment by jkilpatr** [comment]
> I'd say go ahead and merge this, but there are a few issues:
> 
>     * By using block height we make the implementation more complicated and less predictable. Using timestamps would be a superior user experience in most cases, and would simplify the implementation (no need to track block height like you do here). The tradeoff is that we don't know what's involved in spoofing an Ethereum timestamp. We would need to do some more research and come up with a certain buffer time beyond which we don't think a timestamp can be spoofed. Then we would wait through the buffer before freeing a batch on the Cosmos side. I think that it's probably impossible to spoof a timestamp more than a minute, but I have no idea.
> 
>     * In the current implementation, the only way the Cosmos module gains knowledge of the current block height is from an event from Peggy. This multiplies any block time inaccuracies, as well as making it so that batches cannot be freed until a new event has come in. An easy upgrade to this would be to have a process separate from the event oracle that brings in block height. However, someone embarking on this journey would first want to check whether the timestamp route could work.
> 
> 
> Can you make issues for these two upgrades?

issues #169 and #170 made

it is in fact possible to spoof the block time arbitrarily (as far as slowing it down) so no matter what we need to have an oracle event from Ethereum to release funds. Timestamp may be closer to accurate if mining is slow, but if there was a concentrated attempt to spoof on chain clocks timeouts could be delayed indefinitely.
